### ISSUE TITLE
"/products" REST endpoint is duplicating some products and missing others

### ISSUE THREAD

**Issue by MaciekChudek** [woothemes/woocommerce#33386]
When products are listed via the [`/products` REST API endpoint](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products) (and possibly other places), they are by default ordered by `date` `desc`.

In some cases, though, such as following a bulk import, the date will be the same for all products. The problem here is that MySQL does not guarantee stable sorting under these conditions, and so if one was to paginate through the results in small sets it might appear as if products have vanished or have been duplicated. Actually, they are just being ordered differently with each successive query.

A possible way to resolve this would be to always order by `ID` `asc` as the final ordering criteria.

---

**ORIGINAL ISSUE DESCRIPTION:**

"/products" REST endpoint is missing products and replacing them with duplicates of other products.

I have a database of 1833 products. Until recently (possibly due to upgrade to version 6.5.1), the REST API /products endpoint queried them correctly. Now, the returned data contains 54 duplicated products, and a similar number of missing products. 

Browsing the products through the web admin panel shows the correct products (i.e., no duplicates or missing products).

I previously used (but have now deactivated): 
  "WP REST Cache - Must-Use Plugin: by Richard Korthuis - Acato – 2021.3.0". 
I suspect this may have been related, but the issue persists even after deactivating the plugin. I have also tried installing the latest version of this plugin and completely purging it's caches, but the problem persists. 

### Expected behavior

/products endpoint returns one record for each product

### Actual behavior

/products endpoint returns duplicate records for some products but misses others

### Steps to reproduce

```
#!/bin/bash
for PAGE in {1..19}; do	
	curl "https://MY_WEBSITE/wp-json/wc/v3/products?per_page=100&page=$PAGE" \
		-u  MY_CK:MY_CS \
		| sed -z "s|},{|\n},\n{|g" \
		>> temp.txt
done

grep '"id":MISSING_ID' temp.txt | wc -l  # Result: 0
grep '"id":DUPLICATED_ID' temp.txt | wc -l  # Result: 2

```

### WordPress Environment

`

WordPress address (URL): HIDDEN
Site address (URL): HIDDEN
WC Version: 6.5.1
REST API Version: ✔ 6.5.1
WC Blocks Version: ✔ 7.4.3
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.0
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_CA
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.25 (Debian)
PHP Version: 7.3.33-1+0~20211119.91+debian9~1.gbp618351
PHP Post Max Size: 100 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.52.1
OpenSSL/1.0.2u

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.1.48-MariaDB-1~stretch
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.5.1
WC Database Prefix: nos_
Total Database Size: 28.03MB
Database Data Size: 19.57MB
Database Index Size: 8.46MB
nos_woocommerce_sessions: Data: 3.02MB + Index: 0.02MB + Engine InnoDB
nos_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
nos_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
nos_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_actionscheduler_actions: Data: 0.09MB + Index: 0.13MB + Engine InnoDB
nos_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_actionscheduler_logs: Data: 0.08MB + Index: 0.03MB + Engine InnoDB
nos_bv_fw_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_bv_ip_store: Data: 0.33MB + Index: 0.28MB + Engine InnoDB
nos_bv_lp_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
nos_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_options: Data: 4.34MB + Index: 0.39MB + Engine InnoDB
nos_postmeta: Data: 5.41MB + Index: 3.53MB + Engine InnoDB
nos_posts: Data: 2.38MB + Index: 0.81MB + Engine InnoDB
nos_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_term_relationships: Data: 0.23MB + Index: 0.20MB + Engine InnoDB
nos_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
nos_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
nos_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
nos_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
nos_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
nos_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_wc_product_meta_lookup: Data: 0.22MB + Index: 0.45MB + Engine InnoDB
nos_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
nos_woocommerce_square_customers: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
nos_wrc_caches: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
nos_wrc_relations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.13MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_bv_fw_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_bv_ip_store: Data: 0.14MB + Index: 0.09MB + Engine InnoDB
wp_bv_lp_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.52MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB

### Post Type Counts ###

attachment: 2696
nav_menu_item: 10
page: 9
post: 1
product: 1833
revision: 29
shop_order: 9
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 6.5.1

### Inactive Plugins (9) ###

Add From Server: by Dion Hulse – 3.4.5
Better Search Replace: by Delicious Brains – 1.4
Breeze: by Cloudways – 2.0.5
MalCare Security - Free Malware Scanner, Protection & Security for WordPress: by MalCare Security – 4.67
Show Hooks: by rafiq91 – 0.1.6
WooCommerce Square: by WooCommerce – 3.1.0
WP REST Cache - Must-Use Plugin: by Richard Korthuis - Acato – 2021.3.0
WPS Hide Login: by WPServeur
NicolasKulka
wpformation – 1.9.6

WPS Limit Login: by WPServeur
NicolasKulka
wpformation – 1.5.6

### Settings ###

API Enabled: –
Force SSL: –
Currency: CAD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.1.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 218
Oldest: 2022-05-12 01:50:33 -0400
Newest: 2022-06-10 14:01:10 -0400

Pending: 1
Oldest: 2022-06-16 02:00:28 -0400
Newest: 2022-06-16 02:00:28 -0400

### Status report information ###

Generated at: 2022-06-10 14:12:26 -04:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by MaciekChudek** [comment]
Additional debugging info:

Directly querying the database shows the correct data. It's just the REST query that's wrong.

SELECT * FROM `wp_postmeta` WHERE `meta_value`='SKU_OF_ITEM_DUPLICATED'  	# 1 result
SELECT * FROM `wp_postmeta` WHERE `meta_value`='SKU_OF_ITEM_MISSING'		# 1 result

Does the Woo REST API cache product info somewhere rather than querying the database directly? Is there a way to reset that cache?

**Comment by barryhughes** [comment]
@MaciekChudek,

Using the code you shared it would seem possible to confuse `"id"` properties relating to products, with others that relate to term objects (like tags) and perhaps other entities. Even in a modestly sized catalog, "collisions" between the two sets of IDs are not unlikely.

Could this be the cause? Or, have you eliminated this as a possible source of the results you are seeing?

**Comment by MaciekChudek** [comment]
Thanks @barryhughes, interesting insight, but I think I've excluded that. 

By "id", you might mean:

1. The "id" field returned by /products REST queries, which seems to correspond to the wp_posts ID column of the post. I think the uniqueness of these is guaranteed by the MySQL autoincrement schema, but in any case I checked and they are unique.
2. The "sku" that we assigned. I checked and these are also unique. See the SQL queries in my previous comment.

**Comment by barryhughes** [comment]
I mean that the property `id` is used for more than just product IDs. Simplified example of a product object in the response data:

```json
{
    "id": 1234,
    "name": "Some Product",
    "slug": "some-product",
    "permalink": "https:\/\/example.com\/product\/some-product\/",
    "price": "123.45",
    "categories": [
        {
            "id": 800,
            "name": "Foo",
            "slug": "foo"
        },
        {
            "id": 1234,
            "name": "Bar",
            "slug": "bar"
        }
        ],
    "tags": [
        {
            "id": 1235,
            "name": "Baz",
            "slug": "baz"
        }
    ],
    "images": [
        {
            "id": 1240,
            "src": "https:\/\/example.com\/wp-content\/uploads\/2022\/02\/img-foo-bar.png",
            "name": "img-foo-bar.png",
            "alt": ""
        }
    ]
}
```

A term ID (such as for a category or tag) could collide with a post ID, and though image IDs cannot collide with product IDs (both being post IDs) the same image can be reused across multiple products. It doesn't look like the strategy you shared with us eliminates those things, though perhaps you are doing additional work to handle this.

**Comment by barryhughes** [comment]
...If you happen to have the [`jq`](https://stedolan.github.io/jq/) utility available, you could list *only the top-level **(product)** IDs* using something like this:

```sh
jq '.[].id' temp.txt
```

Do you still see duplicates in that list of IDs?

**Comment by MaciekChudek** [comment]
Sorry, I understand now.

Yes, the grep for ids was a shortcut to demonstrate the problem, but I have verified manually by inspecting the file that the product is actually duplicated. For example, in my temp.txt the following line appears exactly identically on lines 201 and 222:

```
{"id":5261,"name":"Leukoplast Sleek LF 1\" Waterproof Tape Pink (12\/bx)","slug":"leukoplast-sleek-lf-1-waterproof-tape-pink-12-bx","permalink":"https:\/\/60medical.ca\/product\/leukoplast-sleek-lf-1-waterproof-tape-pink-12-bx\/","date_created":"2022-05-31T15:48:19","date_created_gmt":"2022-05-31T19:48:19","date_modified":"2022-05-31T15:48:19","date_modified_gmt":"2022-05-31T19:48:19","type":"simple","status":"publish","featured":false,"catalog_visibility":"visible","description":"","short_description":"<p>A reliable and conformable waterproof tape that is easy to unroll\/apply. Resists sweat, grease, oil and water and can stay on for hours or days.<\/p>\n<p>Strong and reliable.<br \/>\nSlight stretch.<br \/>\nMoulds to body contours.<br \/>\nIndication: For fixation of dressing\/tube.<br \/>\nMaterial: Zinc oxide plastic.<\/p>\n","sku":"DSG1037","price":"75","regular_price":"75","sale_price":"","date_on_sale_from":null,"date_on_sale_from_gmt":null,"date_on_sale_to":null,"date_on_sale_to_gmt":null,"on_sale":false,"purchasable":true,"total_sales":0,"virtual":false,"downloadable":false,"downloads":[],"download_limit":-1,"download_expiry":-1,"external_url":"","button_text":"","tax_status":"taxable","tax_class":"","manage_stock":true,"stock_quantity":0,"backorders":"no","backorders_allowed":false,"backordered":false,"low_stock_amount":null,"sold_individually":false,"weight":"","dimensions":{"length":"","width":"","height":""},"shipping_required":true,"shipping_taxable":true,"shipping_class":"","shipping_class_id":0,"reviews_allowed":true,"average_rating":"0.00","rating_count":0,"upsell_ids":[],"cross_sell_ids":[],"parent_id":0,"purchase_note":"","categories":[{"id":30,"name":"DRESSING SUPPLIES","slug":"dressing-supplies"}],"tags":[],"images":[],"attributes":[],"default_attributes":[],"variations":[],"grouped_products":[],"menu_order":0,"price_html":"<span class=\"woocommerce-Price-amount amount\"><bdi><span class=\"woocommerce-Price-currencySymbol\">&#36;<\/span>75.00<\/bdi><\/span>","related_ids":[248,192,201,268,148],"meta_data":[],"stock_status":"outofstock","has_options":false,"_links":{"self":[{"href":"https:\/\/60medical.ca\/wp-json\/wc\/v3\/products\/5261"}],"collection":[{"href":"https:\/\/60medical.ca\/wp-json\/wc\/v3\/products"}]}
},
```

**Comment by barryhughes** [comment]
Thanks ... though I cannot replicate locally.

The product objects themselves will be cached if possible, but the query to determine which product IDs to retrieve is not cached. 

I realize you have already tried to eliminate the potential for conflicts but I'm curious: if you create a brand new WordPress site, installing and activating only WooCommerce, and import a database dump from your existing site: is the same thing replicable there?

**Comment by MaciekChudek** [comment]
Thanks. I'm just boarding a flight. I'll attempt this tomorrow and report
back.

**Comment by pierrehenrinortje** [comment]
I understand that this issue was automatically closed, but I am getting the same issue on a site I'm working on.
The site is populated with over 100 products, which requires "page" and "per_page" parameters to be used when querying, combined with a loop to get all the products. Not only am I missing products, I am also getting duplicates when comparing the results of each page. Furthermore, these duplicates and missing products are not tied to specific products, adjusting the parameters results in different records duplicating/missing.

Woocommerce Version: 6.8.2
Wordpress Version: 6.0.2
Woocommerce Database Version: 6.6.1

I'm running the queries in Golang, but I've also confirmed the bug using Postman, using different parameter values, I get the same behavior. I have confirmed that the duplicate products do not belong to separate products, only one. I've also confirmed that the products missing from the result set can be found when querying with their respective product ID's (when getting a single product), or when changing the "page", "per_page" parameter values.

Please let me know if this bug is still relevant, thanks.

**Comment by MaciekChudek** [comment]
To clarify, I stopped responding to the original issue because I came down
with a major medical condition, not because it was fixed. Better now, so
I'll resume trying to fix it by migration to a new WordPress insurance in
the next few days.

**Comment by barryhughes** [comment]
> To clarify, I stopped responding to the original issue because I came down with a major medical condition, not because it was fixed.

I'm sorry to hear that.

Our assumption was not that the problem was fixed, though the wording in the automated actions may have given that impression.

> Please let me know if this bug is still relevant, thanks.

The biggest problem here from our side is our inability to reproduce. Assuming the data is stable (by which I mean, products are not being created, updated or deleted during pagination) and assuming there are no caching effects outside of WooCommerce's control, then I wouldn't expect to see this problem.

What might be valuable is enabling [query logging](https://dev.mysql.com/doc/refman/5.7/en/query-log.html) and matching the problematic REST API responses to the captured queries. Things to look for might include:

- Are there cases where no query is logged (this could tell us the problem exists within a layer of caching)?
- Are the queries 'correct' (could they have been modified by a third-party plugin or other component)?
- Do the query results and REST API responses convey essentially the same data (again, I'm interested in the potential for results to be modified post-query)?

Another potentially useful thing could be if a non-production sandbox or testing environment (where the problem can easily be seen) can be made available, and if we can be permitted access to troubleshoot.

I can definitely re-open this, but we still need more data if we are to confirm the bug and identify its source.

**Comment by pierrehenrinortje** [comment]
ok thanks @barryhughes , I will run through the sanity checks and confirm with you soon.

**Comment by pierrehenrinortje** [comment]
@barryhughes @MaciekChudek I was able to identify the issue:
All the products that I added to the woocommerce site were created using a batch update, 100 products added per batch at most.
Each batch of products share the same date value (created date), which happens to be the default "orderby" parameter value set when listing all products. This results in a bad scenario where ordering by date with a smaller per_page range results in erroneous result sets.

Once I changed the "orderby" parameter to "id", the problem was solved and the result sets no longer contain duplicates or missing records.

**Comment by barryhughes** [comment]
Interesting. This makes me wonder if we should always order by ID as a final sorting criteria (that way the primary order-by criteria will be respected, but we will also have stable sorting).

**Comment by barryhughes** [comment]
I've updated the description and this has been added to our backlog.

**Comment by alexzk1** [comment]
Can confirm. I do java program which should bulk pull items from DB for the customer (I have no details about woo used, just i know V3 API).
Woo responds with duplicates, page size 100 gives roughly 1.5x times less duplicates then page size 50. And missing some values of coz, i.e. values are replaced by dups of other values.

**Comment by jelledruut** [comment]
Experiencing the same. A workaround I found: add the following query parameter:
`"orderby": "id"`

**Comment by drivs** [comment]
> Interesting. This makes me wonder if we should always order by ID as a final sorting criteria (that way the primary order-by criteria will be respected, but we will also have stable sorting).

Hi there,

For me it is still missing around ~40 products out of ~1900. Looking them up, they are all published at the same time.

I used the query parameters `order_by' = id`, and it is still not returning all the products. This is quite frustrating, since the products on WooCommerce are periodically imported from an external system, and therefore always have the same publish date.
I hope you will look into fixing this issue.

Best regards
Andreas

**Comment by MaciekChudek** [comment]
It's been a while, but I just want to confirm that adding param "orderby":"id" to my API call fixes the issue for me. I also want to confirm that the affected items did have identical creation times due to batch creation/update.

I believe @barryhughes diagnosis of the cause (pagination with unstable ordering of simultaneously dated items) is correct.

Not sure what the etiquette is. Should I close this issue, or only once #34546 is closed?

**Comment by barryhughes** [comment]
I think leave both open. #33386 covers product queries, and #34546 covers order queries. They will be closed once PRs are accepted and merged in each case.

**Comment by VanShane** [comment]
Hi, I'm encoutering the same issue as described above. 

We have about 188 products, when consulting the api `/v3/products`, we do get 188 products between two pages with `page_limit` at 100. But there are 50 duplicates and so the same amount missing. I've checked the creation dates and all of them are the same. When I add `order_by=id` in the api querry I still get the same issue, the same products are being duplicated.

Has anyone found a solution to this problem ?

Thanks in advance.

**Comment by barryhughes** [comment]
> When I add order_by=id in the api querry I still get the same issue

If you are literally using `order_by`, I think that is the wrong parameter name. It should be `orderby`. 

📝 [woocommerce.github.io/woocommerce-rest-api-docs/#list-all-orders](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-orders)

Or, if you are already doing that, can you confirm if it is being respected at all (is each page ordered by ID, even if 'duplication' is happening across pages)?

**Comment by VanShane** [comment]
> > When I add order_by=id in the api querry I still get the same issue
> 
> If you are literally using `order_by`, I think that is the wrong parameter name. It should be `orderby`.
> 
> 📝 [woocommerce.github.io/woocommerce-rest-api-docs/#list-all-orders](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-orders)
> 
> Or, if you are already doing that, can you confirm if it is being respected at all (is each page ordered by ID, even if 'duplication' is happening across pages)?

Hi @barryhughes,

Thanks for the info, my bad. It does work with `orderby` indeed ! I can now retrieve all my products, I have no more duplicates and they are in fact ordered by id.

Thanks for the help !

**Comment by barryhughes** [comment]
That's great to hear (also because it's further confirmation that ordering by ID will help alleviate this problem, at least in some cases).

**Comment by eyepokes** [comment]
It's good to hear that you found a workaround with orderby=id, but why is everyone acting like this is normal behavior for an important part of the API? The problem has existed for 2 years and is still not resolved. I'm using Woo 8.7.0 and last night I thought I was crazy until I realized I ran into this problem. To reproduce the error, simply add more than 1 thousand products to different categories via the API and make a request to the /products method for a category that has more than 100 products.

**Comment by agusgrana** [comment]
This behaviour is unusual for an API, it should be patched so that we don't have to use alternative ways to get the data from the API. When I create an API in node, if I get this error I try to fix it. Because the API request is not implemented by default to sort the result by Id instead of sorting by creation date. 

Best regards

**Comment by droidsheep86** [comment]
I have the same problem and none of the solution here is helping. For us the situation is even worse. We have around 2100 products, but rest api shows only 980. Also lots of duplicates, lots of missing products.

**Comment by barryhughes** [comment]
Even if the symptoms are somewhat similar, I'm not sure that is the same problem *(especially* if the shared solutions don't work for you) 🤔 

If you look at the `X-WP-Total` header, do you only see 980?

```sh
wget -o- --server-response --spider \
    --header="Authorization: Basic your-encoded-key" \
    https://example/wp-json/wc/v3/products | grep "X-WP-Total"
```

If so, I'd recommend creating a new issue.

**Comment by droidsheep86** [comment]
Sorry for the late response. But in our case it was something else, changed pre_get_posts filter. Ordering by ID as proposed here works just fine. Sorry for the confusion.

**Comment by barryhughes** [comment]
Np, appreciate the clarification. It's also a good call-out, as I think perhaps a few folks have encountered what appears to be the same problem, except ordering by ID has *not* worked (in which case, it could be a similar problem to the one you faced).

Was it custom code/a third party plugin that was hooked into `pre_get_posts` causing the issue (if you are happy to share)?

**Comment by rwkyyy** [comment]
Can confirm, in a total of 10633 SKU's, it returned 134 duplicates!

Used the following parameters:

- _per_page_
- _page_

Returned **134 duplicates** (sku checking)

Used the following parameters:

- _per_page_
- _offset_

Returned **184 duplicates** (sku checking).

A friend of mine @MihaiDumitru97, tested in Python and got 186 duplicates, this is how I found out of the issue.

This is very inconsistent and unreliable!

**Solution:** _As a workaround, I used the ordering by ID and ASC, this helped in clearing out duplicates, but it should either be mentioned in the docs or have this applied by default._

p.s. latest WP & WC

**Comment by rpldom** [comment]
I also have this issue with duplicate products. It's absolutely inappropriate for such major API endpoint to have this flaw.

**Comment by chubes4** [comment]
This looks fixed on current trunk by #61410 / commit ce69a6b95e29eeccf6d78042f2356ac0acdb5e58.

Source check:

- `/wc/v3/products` is handled by `WC_REST_Products_Controller`, which extends `WC_REST_Products_V2_Controller`.
- `WC_REST_Products_V2_Controller` extends `WC_REST_CRUD_Controller`.
- `WC_REST_CRUD_Controller::get_collection_params()` still defaults collections to `orderby=date` and `order=desc`.
- `WC_REST_CRUD_Controller::prepare_objects_query()` now appends `ID` as a secondary sort whenever `orderby` is `date` or `modified`:

```php
// Add ID to orderby for consistency with pagination.
if ( in_array( $args['orderby'], array( 'date', 'modified' ), true ) ) {
    $args['orderby'] = $args['orderby'] . ' ID';
}
```

That means the default product-list query now becomes a stable date + ID ordering, which is the proposed fix in this issue. PR #61410 explicitly mentioned this issue URL in its description, but did not use GitHub auto-close syntax, so this issue appears to have remained open after the shared REST CRUD fix landed.

If there are still duplicates/missing products on a build that includes ce69a6b95e or later, it would be useful to capture the exact generated SQL and confirm whether the query includes the ID tie-breaker. Otherwise I think this can be closed as fixed by #61410.

**Comment by jamesckemp** [comment]
Hi @MaciekChudek - thanks for reporting this, and for the detail you gave.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

It looks like this has already been fixed, or it's no longer happening on trunk.

Here's what I found: Current trunk `WC_REST_CRUD_Controller::prepare_objects_query()` now appends `ID` as a secondary sort key whenever collection ordering uses `date` or `modified`. That is the stable-pagination fix the issue discussion converged on for duplicate/missing `/products` REST results caused by identical timestamps.

I'm going to close this one. If you're still facing this issue on a current version, please reply with steps to reproduce and we'll happily take another look.

Fix/changelog reference: https://github.com/woocommerce/woocommerce/pull/61410
